File: frameworks/mobile/android-compose/DataTableComplex.kt
package com.accessibility.benchmark.compose

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class TableRow(
    val id: String,
    val name: String,
    val email: String,
    val department: String,
    val salary: Double,
    val status: String,
    val joinDate: String,
    val isSelected: Boolean = false
)

@Composable
fun DataTableComplex() {
    var searchQuery by remember { mutableStateOf("") }
    var selectedDepartment by remember { mutableStateOf("All") }
    var sortColumn by remember { mutableStateOf("name") }
    var sortAscending by remember { mutableStateOf(true) }
    var currentPage by remember { mutableStateOf(1) }
    var showFilters by remember { mutableStateOf(false) }
    var selectedRows by remember { mutableStateOf(setOf<String>()) }
    var showBulkActions by remember { mutableStateOf(false) }
    
    val departments = listOf("All", "Engineering", "Marketing", "Sales", "HR", "Finance")
    val itemsPerPage = 10
    
    val sampleData = remember {
        listOf(
            TableRow("1", "John Doe", "john@example.com", "Engineering", 75000.0, "Active", "2023-01-15"),
            TableRow("2", "Jane Smith", "jane@example.com", "Marketing", 65000.0, "Active", "2023-02-20"),
            TableRow("3", "Mike Johnson", "mike@example.com", "Sales", 70000.0, "Inactive", "2023-03-10"),
            TableRow("4", "Sarah Wilson", "sarah@example.com", "HR", 60000.0, "Active", "2023-04-05"),
            TableRow("5", "David Brown", "david@example.com", "Finance", 80000.0, "Active", "2023-05-12"),
            TableRow("6", "Lisa Davis", "lisa@example.com", "Engineering", 72000.0, "Active", "2023-06-18"),
            TableRow("7", "Tom Miller", "tom@example.com", "Marketing", 68000.0, "Inactive", "2023-07-22"),
            TableRow("8", "Amy Garcia", "amy@example.com", "Sales", 69000.0, "Active", "2023-08-30"),
            TableRow("9", "Chris Lee", "chris@example.com", "HR", 62000.0, "Active", "2023-09-14"),
            TableRow("10", "Emma Taylor", "emma@example.com", "Finance", 78000.0, "Active", "2023-10-08")
        )
    }
    
    val filteredData = remember(sampleData, searchQuery, selectedDepartment) {
        sampleData.filter { row ->
            val matchesSearch = row.name.contains(searchQuery, ignoreCase = true) ||
                    row.email.contains(searchQuery, ignoreCase = true)
            val matchesDepartment = selectedDepartment == "All" || row.department == selectedDepartment
            matchesSearch && matchesDepartment
        }
    }
    
    val sortedData = remember(filteredData, sortColumn, sortAscending) {
        filteredData.sortedWith { a, b ->
            val comparison = when (sortColumn) {
                "name" -> a.name.compareTo(b.name)
                "email" -> a.email.compareTo(b.email)
                "department" -> a.department.compareTo(b.department)
                "salary" -> a.salary.compareTo(b.salary)
                "status" -> a.status.compareTo(b.status)
                "joinDate" -> a.joinDate.compareTo(b.joinDate)
                else -> 0
            }
            if (sortAscending) comparison else -comparison
        }
    }
    
    val totalPages = (sortedData.size + itemsPerPage - 1) / itemsPerPage
    val startIndex = (currentPage - 1) * itemsPerPage
    val endIndex = minOf(startIndex + itemsPerPage, sortedData.size)
    val pageData = sortedData.subList(startIndex, endIndex)
    
    Column(
        modifier = Modifier
            .fillMaxSize()
            .padding(16.dp)
    ) {
        // Header
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Employee Data Table",
                fontSize = 24.sp,
                fontWeight = FontWeight.Bold
            )
            
            Row {
                IconButton(onClick = { showFilters = !showFilters }) {
                    Icon(Icons.Default.FilterList, contentDescription = null)
                }
                IconButton(onClick = { /* Export functionality */ }) {
                    Icon(Icons.Default.FileDownload, contentDescription = null)
                }
            }
        }
        
        Spacer(modifier = Modifier.height(16.dp))
        
        // Search and Filter Bar
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(8.dp)
        ) {
            OutlinedTextField(
                value = searchQuery,
                onValueChange = { searchQuery = it },
                modifier = Modifier.weight(1f),
                placeholder = { Text("Search employees...") },
                leadingIcon = { Icon(Icons.Default.Search, contentDescription = null) }
            )
            
            DropdownMenu(
                expanded = showFilters,
                onDismissRequest = { showFilters = false }
            ) {
                departments.forEach { dept ->
                    DropdownMenuItem(onClick = {
                        selectedDepartment = dept
                        showFilters = false
                    }) {
                        Text(dept)
                    }
                }
            }
            
            Button(
                onClick = { showFilters = true },
                modifier = Modifier.padding(horizontal = 8.dp)
            ) {
                Text("Filter")
                Icon(Icons.Default.ArrowDropDown, contentDescription = null)
            }
        }
        
        Spacer(modifier = Modifier.height(16.dp))
        
        // Bulk Actions
        if (selectedRows.isNotEmpty()) {
            Card(
                modifier = Modifier.fillMaxWidth(),
                backgroundColor = MaterialTheme.colors.primary.copy(alpha = 0.1f)
            ) {
                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = "${selectedRows.size} items selected",
                        fontWeight = FontWeight.Medium
                    )
                    
                    Row {
                        Button(onClick = { /* Bulk edit */ }) {
                            Text("Edit")
                        }
                        Spacer(modifier = Modifier.width(8.dp))
                        Button(onClick = { /* Bulk delete */ }) {
                            Text("Delete")
                        }
                        Spacer(modifier = Modifier.width(8.dp))
                        Button(onClick = { /* Bulk export */ }) {
                            Text("Export")
                        }
                    }
                }
            }
            
            Spacer(modifier = Modifier.height(8.dp))
        }
        
        // Table Header
        Card(
            modifier = Modifier.fillMaxWidth(),
            elevation = 2.dp
        ) {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .background(MaterialTheme.colors.primary.copy(alpha = 0.1f))
                    .padding(16.dp),
                horizontalArrangement = Arrangement.SpaceBetween
            ) {
                Row(
                    modifier = Modifier.weight(0.1f),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Checkbox(
                        checked = selectedRows.size == pageData.size && pageData.isNotEmpty(),
                        onCheckedChange = { checked ->
                            if (checked) {
                                selectedRows = pageData.map { it.id }.toSet()
                            } else {
                                selectedRows = emptySet()
                            }
                        }
                    )
                }
                
                TableHeaderCell(
                    text = "Name",
                    weight = 0.2f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { sortColumn = "name"; sortAscending = !sortAscending }
                )
                
                TableHeaderCell(
                    text = "Email",
                    weight = 0.25f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { sortColumn = "email"; sortAscending = !sortAscending }
                )
                
                TableHeaderCell(
                    text = "Department",
                    weight = 0.15f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { sortColumn = "department"; sortAscending = !sortAscending }
                )
                
                TableHeaderCell(
                    text = "Salary",
                    weight = 0.1f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { sortColumn = "salary"; sortAscending = !sortAscending }
                )
                
                TableHeaderCell(
                    text = "Status",
                    weight = 0.1f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { sortColumn = "status"; sortAscending = !sortAscending }
                )
                
                TableHeaderCell(
                    text = "Actions",
                    weight = 0.1f,
                    sortColumn = sortColumn,
                    currentSort = sortColumn,
                    sortAscending = sortAscending,
                    onSort = { /* No sort for actions */ }
                )
            }
        }
        
        Spacer(modifier = Modifier.height(8.dp))
        
        // Table Body
        LazyColumn(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(4.dp)
        ) {
            items(pageData) { row ->
                TableRowItem(
                    row = row,
                    isSelected = selectedRows.contains(row.id),
                    onSelectionChange = { selected ->
                        if (selected) {
                            selectedRows = selectedRows + row.id
                        } else {
                            selectedRows = selectedRows - row.id
                        }
                    },
                    onEdit = { /* Edit row */ },
                    onDelete = { /* Delete row */ },
                    onView = { /* View details */ }
                )
            }
        }
        
        Spacer(modifier = Modifier.height(16.dp))
        
        // Pagination
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Showing ${startIndex + 1}-${endIndex} of ${sortedData.size} results",
                fontSize = 14.sp,
                color = MaterialTheme.colors.onSurface.copy(alpha = 0.6f)
            )
            
            Row(
                horizontalArrangement = Arrangement.spacedBy(8.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Button(
                    onClick = { currentPage = maxOf(1, currentPage - 1) },
                    enabled = currentPage > 1
                ) {
                    Text("Previous")
                }
                
                Text(
                    text = "Page $currentPage of $totalPages",
                    fontSize = 14.sp
                )
                
                Button(
                    onClick = { currentPage = minOf(totalPages, currentPage + 1) },
                    enabled = currentPage < totalPages
                ) {
                    Text("Next")
                }
            }
        }
    }
}

@Composable
fun TableHeaderCell(
    text: String,
    weight: Float,
    sortColumn: String,
    currentSort: String,
    sortAscending: Boolean,
    onSort: () -> Unit
) {
    Row(
        modifier = Modifier
            .weight(weight)
            .clickable { onSort() },
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            text = text,
            fontWeight = FontWeight.Bold,
            fontSize = 14.sp
        )
        
        if (sortColumn == currentSort) {
            Icon(
                imageVector = if (sortAscending) Icons.Default.ArrowUpward else Icons.Default.ArrowDownward,
                contentDescription = null,
                modifier = Modifier.size(16.dp)
            )
        }
    }
}

@Composable
fun TableRowItem(
    row: TableRow,
    isSelected: Boolean,
    onSelectionChange: (Boolean) -> Unit,
    onEdit: () -> Unit,
    onDelete: () -> Unit,
    onView: () -> Unit
) {
    Card(
        modifier = Modifier.fillMaxWidth(),
        elevation = 1.dp,
        backgroundColor = if (isSelected) MaterialTheme.colors.primary.copy(alpha = 0.1f) else MaterialTheme.colors.surface
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Row(
                modifier = Modifier.weight(0.1f),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Checkbox(
                    checked = isSelected,
                    onCheckedChange = onSelectionChange
                )
            }
            
            Text(
                text = row.name,
                modifier = Modifier.weight(0.2f),
                fontSize = 14.sp
            )
            
            Text(
                text = row.email,
                modifier = Modifier.weight(0.25f),
                fontSize = 14.sp,
                color = MaterialTheme.colors.primary
            )
            
            Text(
                text = row.department,
                modifier = Modifier.weight(0.15f),
                fontSize = 14.sp
            )
            
            Text(
                text = "$${String.format("%.0f", row.salary)}",
                modifier = Modifier.weight(0.1f),
                fontSize = 14.sp,
                textAlign = TextAlign.End
            )
            
            Text(
                text = row.status,
                modifier = Modifier.weight(0.1f),
                fontSize = 14.sp,
                color = if (row.status == "Active") Color.Green else Color.Red
            )
            
            Row(
                modifier = Modifier.weight(0.1f),
                horizontalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                IconButton(onClick = onView) {
                    Icon(Icons.Default.Visibility, contentDescription = null, modifier = Modifier.size(16.dp))
                }
                IconButton(onClick = onEdit) {
                    Icon(Icons.Default.Edit, contentDescription = null, modifier = Modifier.size(16.dp))
                }
                IconButton(onClick = onDelete) {
                    Icon(Icons.Default.Delete, contentDescription = null, modifier = Modifier.size(16.dp))
                }
            }
        }
    }
}

File: frameworks/mobile/flutter/lib/data_table_complex.dart
import 'package:flutter/material.dart';

class DataTableComplex extends StatefulWidget {
  const DataTableComplex({Key? key}) : super(key: key);

  @override
  State<DataTableComplex> createState() => _DataTableComplexState();
}

class _DataTableComplexState extends State<DataTableComplex> {
  String searchQuery = '';
  String sortColumn = 'name';
  bool sortAscending = true;
  Set<String> selectedRows = {};
  int currentPage = 1;
  bool showFilters = false;
  String departmentFilter = 'All';
  String statusFilter = 'All';
  int itemsPerPage = 10;
  bool isLoading = false;

  final List<String> departments = ['All', 'Engineering', 'Marketing', 'Sales', 'HR', 'Finance'];
  final List<String> statuses = ['All', 'Active', 'Inactive', 'Pending', 'Terminated'];

  final List<Map<String, dynamic>> sampleData = [
    {
      'id': '1',
      'name': 'John Doe',
      'email': 'john@example.com',
      'department': 'Engineering',
      'salary': 75000.0,
      'status': 'Active',
      'joinDate': '2023-01-15',
    },
    {
      'id': '2',
      'name': 'Jane Smith',
      'email': 'jane@example.com',
      'department': 'Marketing',
      'salary': 65000.0,
      'status': 'Active',
      'joinDate': '2023-02-20',
    },
    {
      'id': '3',
      'name': 'Mike Johnson',
      'email': 'mike@example.com',
      'department': 'Sales',
      'salary': 70000.0,
      'status': 'Inactive',
      'joinDate': '2023-03-10',
    },
    {
      'id': '4',
      'name': 'Sarah Wilson',
      'email': 'sarah@example.com',
      'department': 'HR',
      'salary': 60000.0,
      'status': 'Active',
      'joinDate': '2023-04-05',
    },
    {
      'id': '5',
      'name': 'David Brown',
      'email': 'david@example.com',
      'department': 'Finance',
      'salary': 80000.0,
      'status': 'Active',
      'joinDate': '2023-05-12',
    },
    {
      'id': '6',
      'name': 'Lisa Davis',
      'email': 'lisa@example.com',
      'department': 'Engineering',
      'salary': 72000.0,
      'status': 'Active',
      'joinDate': '2023-06-18',
    },
    {
      'id': '7',
      'name': 'Tom Miller',
      'email': 'tom@example.com',
      'department': 'Marketing',
      'salary': 68000.0,
      'status': 'Inactive',
      'joinDate': '2023-07-22',
    },
    {
      'id': '8',
      'name': 'Amy Garcia',
      'email': 'amy@example.com',
      'department': 'Sales',
      'salary': 69000.0,
      'status': 'Active',
      'joinDate': '2023-08-30',
    },
    {
      'id': '9',
      'name': 'Chris Lee',
      'email': 'chris@example.com',
      'department': 'HR',
      'salary': 62000.0,
      'status': 'Active',
      'joinDate': '2023-09-14',
    },
    {
      'id': '10',
      'name': 'Emma Taylor',
      'email': 'emma@example.com',
      'department': 'Finance',
      'salary': 78000.0,
      'status': 'Active',
      'joinDate': '2023-10-08',
    },
  ];

  List<Map<String, dynamic>> get filteredData {
    return sampleData.where((row) {
      final matchesSearch = row['name'].toString().toLowerCase().contains(searchQuery.toLowerCase()) ||
                           row['email'].toString().toLowerCase().contains(searchQuery.toLowerCase());
      final matchesDepartment = departmentFilter == 'All' || row['department'] == departmentFilter;
      final matchesStatus = statusFilter == 'All' || row['status'] == statusFilter;
      return matchesSearch && matchesDepartment && matchesStatus;
    }).toList();
  }

  List<Map<String, dynamic>> get sortedData {
    final filtered = filteredData;
    filtered.sort((a, b) {
      dynamic aValue = a[sortColumn];
      dynamic bValue = b[sortColumn];
      
      if (aValue is String && bValue is String) {
        return sortAscending ? aValue.compareTo(bValue) : bValue.compareTo(aValue);
      } else if (aValue is num && bValue is num) {
        return sortAscending ? aValue.compareTo(bValue) : bValue.compareTo(aValue);
      }
      return 0;
    });
    return filtered;
  }

  int get totalPages => (sortedData.length / itemsPerPage).ceil();
  int get startIndex => (currentPage - 1) * itemsPerPage;
  int get endIndex => (startIndex + itemsPerPage).clamp(0, sortedData.length);
  List<Map<String, dynamic>> get pageData => sortedData.sublist(startIndex, endIndex);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.grey[100],
      appBar: AppBar(
        title: const Text('Employee Data Table'),
        backgroundColor: Colors.white,
        foregroundColor: Colors.black,
        elevation: 1,
        actions: [
          IconButton(
            icon: const Icon(Icons.filter_list),
            onPressed: () => setState(() => showFilters = !showFilters),
          ),
          IconButton(
            icon: const Icon(Icons.download),
            onPressed: () => _exportData('CSV'),
          ),
          IconButton(
            icon: const Icon(Icons.refresh),
            onPressed: _loadData,
          ),
        ],
      ),
      body: Column(
        children: [
          _buildSearchBar(),
          if (showFilters) _buildFilters(),
          if (selectedRows.isNotEmpty) _buildBulkActions(),
          Expanded(
            child: isLoading
                ? const Center(child: CircularProgressIndicator())
                : sortedData.isEmpty
                    ? _buildEmptyState()
                    : _buildDataTable(),
          ),
          _buildPagination(),
        ],
      ),
    );
  }

  Widget _buildSearchBar() {
    return Container(
      padding: const EdgeInsets.all(16),
      color: Colors.white,
      child: Row(
        children: [
          Expanded(
            child: TextField(
              decoration: const InputDecoration(
                hintText: 'Search employees...',
                prefixIcon: Icon(Icons.search),
                border: OutlineInputBorder(),
                contentPadding: EdgeInsets.symmetric(horizontal: 12, vertical: 8),
              ),
              onChanged: (value) => setState(() => searchQuery = value),
            ),
          ),
          const SizedBox(width: 12),
          ElevatedButton.icon(
            onPressed: () => setState(() => showFilters = !showFilters),
            icon: const Icon(Icons.filter_list),
            label: const Text('Filter'),
          ),
        ],
      ),
    );
  }

  Widget _buildFilters() {
    return Container(
      padding: const EdgeInsets.all(16),
      color: Colors.white,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              const Text('Filters', style: TextStyle(fontSize: 18, fontWeight: FontWeight.bold)),
              TextButton(
                onPressed: _clearFilters,
                child: const Text('Clear All'),
              ),
            ],
          ),
          const SizedBox(height: 16),
          Row(
            children: [
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text('Department', style: TextStyle(fontWeight: FontWeight.w500)),
                    const SizedBox(height: 8),
                    DropdownButtonFormField<String>(
                      value: departmentFilter,
                      decoration: const InputDecoration(
                        border: OutlineInputBorder(),
                        contentPadding: EdgeInsets.symmetric(horizontal: 12, vertical: 8),
                      ),
                      items: departments.map((dept) => DropdownMenuItem(
                        value: dept,
                        child: Text(dept),
                      )).toList(),
                      onChanged: (value) => setState(() => departmentFilter = value!),
                    ),
                  ],
                ),
              ),
              const SizedBox(width: 16),
              Expanded(
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    const Text('Status', style: TextStyle(fontWeight: FontWeight.w500)),
                    const SizedBox(height: 8),
                    DropdownButtonFormField<String>(
                      value: statusFilter,
                      decoration: const InputDecoration(
                        border: OutlineInputBorder(),
                        contentPadding: EdgeInsets.symmetric(horizontal: 12, vertical: 8),
                      ),
                      items: statuses.map((status) => DropdownMenuItem(
                        value: status,
                        child: Text(status),
                      )).toList(),
                      onChanged: (value) => setState(() => statusFilter = value!),
                    ),
                  ],
                ),
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildBulkActions() {
    return Container(
      padding: const EdgeInsets.all(16),
      color: Colors.blue[50],
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Text(
            '${selectedRows.length} item${selectedRows.length != 1 ? 's' : ''} selected',
            style: const TextStyle(fontWeight: FontWeight.w600, color: Colors.blue),
          ),
          Row(
            children: [
              TextButton(
                onPressed: () => _handleBulkAction('Edit'),
                child: const Text('Edit'),
              ),
              const SizedBox(width: 8),
              TextButton(
                onPressed: () => _handleBulkAction('Delete'),
                child: const Text('Delete', style: TextStyle(color: Colors.red)),
              ),
              const SizedBox(width: 8),
              TextButton(
                onPressed: () => _handleBulkAction('Export'),
                child: const Text('Export'),
              ),
            ],
          ),
        ],
      ),
    );
  }

  Widget _buildDataTable() {
    return Card(
      margin: const EdgeInsets.all(16),
      child: SingleChildScrollView(
        scrollDirection: Axis.horizontal,
        child: DataTable(
          columns: [
            DataColumn(
              label: Checkbox(
                value: selectedRows.length == pageData.length && pageData.isNotEmpty,
                onChanged: _handleSelectAll,
              ),
            ),
            DataColumn(
              label: _buildSortableHeader('Name', 'name'),
              onSort: (columnIndex, ascending) => _handleSort('name'),
            ),
            DataColumn(
              label: _buildSortableHeader('Email', 'email'),
              onSort: (columnIndex, ascending) => _handleSort('email'),
            ),
            DataColumn(
              label: _buildSortableHeader('Department', 'department'),
              onSort: (columnIndex, ascending) => _handleSort('department'),
            ),
            DataColumn(
              label: _buildSortableHeader('Salary', 'salary'),
              onSort: (columnIndex, ascending) => _handleSort('salary'),
            ),
            DataColumn(
              label: _buildSortableHeader('Status', 'status'),
              onSort: (columnIndex, ascending) => _handleSort('status'),
            ),
            const DataColumn(label: Text('Actions')),
          ],
          rows: pageData.map((row) => DataRow(
            selected: selectedRows.contains(row['id']),
            onSelectChanged: (selected) => _handleSelectRow(row['id']),
            cells: [
              DataCell(Checkbox(
                value: selectedRows.contains(row['id']),
                onChanged: (selected) => _handleSelectRow(row['id']),
              )),
              DataCell(Text(row['name'])),
              DataCell(Text(row['email'], style: const TextStyle(color: Colors.blue))),
              DataCell(Text(row['department'])),
              DataCell(Text('\$${row['salary'].toStringAsFixed(0)}')),
              DataCell(
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 8, vertical: 4),
                  decoration: BoxDecoration(
                    color: row['status'] == 'Active' ? Colors.green : Colors.red,
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: Text(
                    row['status'],
                    style: const TextStyle(color: Colors.white, fontSize: 12),
                  ),
                ),
              ),
              DataCell(Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  IconButton(
                    icon: const Icon(Icons.visibility, size: 16),
                    onPressed: () => _handleRowAction('view', row['id']),
                  ),
                  IconButton(
                    icon: const Icon(Icons.edit, size: 16),
                    onPressed: () => _handleRowAction('edit', row['id']),
                  ),
                  IconButton(
                    icon: const Icon(Icons.delete, size: 16, color: Colors.red),
                    onPressed: () => _handleRowAction('delete', row['id']),
                  ),
                ],
              )),
            ],
          )).toList(),
        ),
      ),
    );
  }

  Widget _buildSortableHeader(String title, String column) {
    return Row(
      mainAxisSize: MainAxisSize.min,
      children: [
        Text(title, style: const TextStyle(fontWeight: FontWeight.bold)),
        if (sortColumn == column)
          Icon(
            sortAscending ? Icons.arrow_upward : Icons.arrow_downward,
            size: 16,
          )
        else
          const Icon(Icons.arrow_upward, size: 16, color: Colors.grey),
      ],
    );
  }

  Widget _buildEmptyState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(Icons.table_chart, size: 64, color: Colors.grey[400]),
          const SizedBox(height: 16),
          const Text(
            'No data found',
            style: TextStyle(fontSize: 24, fontWeight: FontWeight.bold),
          ),
          const SizedBox(height: 8),
          Text(
            'Try adjusting your search or filter criteria',
            style: TextStyle(color: Colors.grey[600]),
          ),
          const SizedBox(height: 16),
          ElevatedButton(
            onPressed: _clearFilters,
            child: const Text('Clear all filters'),
          ),
        ],
      ),
    );
  }

  Widget _buildPagination() {
    return Container(
      padding: const EdgeInsets.all(16),
      color: Colors.white,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Text(
            'Showing ${startIndex + 1}-${endIndex} of ${sortedData.length} results',
            style: TextStyle(color: Colors.grey[600]),
          ),
          Row(
            children: [
              IconButton(
                onPressed: currentPage > 1 ? () => setState(() => currentPage--) : null,
                icon: const Icon(Icons.chevron_left),
              ),
              Text('Page $currentPage of $totalPages'),
              IconButton(
                onPressed: currentPage < totalPages ? () => setState(() => currentPage++) : null,
                icon: const Icon(Icons.chevron_right),
              ),
              const SizedBox(width: 16),
              DropdownButton<int>(
                value: itemsPerPage,
                items: [5, 10, 25, 50, 100].map((value) => DropdownMenuItem(
                  value: value,
                  child: Text('$value per page'),
                )).toList(),
                onChanged: (value) => setState(() {
                  itemsPerPage = value!;
                  currentPage = 1;
                }),
              ),
            ],
          ),
        ],
      ),
    );
  }

  void _handleSort(String column) {
    setState(() {
      if (sortColumn == column) {
        sortAscending = !sortAscending;
      } else {
        sortColumn = column;
        sortAscending = true;
      }
    });
  }

  void _handleSelectAll(bool? selected) {
    setState(() {
      if (selected == true) {
        selectedRows = pageData.map((row) => row['id'] as String).toSet();
      } else {
        selectedRows.clear();
      }
    });
  }

  void _handleSelectRow(String id) {
    setState(() {
      if (selectedRows.contains(id)) {
        selectedRows.remove(id);
      } else {
        selectedRows.add(id);
      }
    });
  }

  void _handleBulkAction(String action) {
    if (selectedRows.isEmpty) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(content: Text('Please select items to perform bulk actions')),
      );
      return;
    }
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Bulk action: $action on ${selectedRows.length} selected rows')),
    );
  }

  void _handleRowAction(String action, String id) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Row action: $action on row $id')),
    );
  }

  void _exportData(String format) {
    ScaffoldMessenger.of(context).showSnackBar(
      SnackBar(content: Text('Exporting data as $format')),
    );
  }

  void _clearFilters() {
    setState(() {
      searchQuery = '';
      departmentFilter = 'All';
      statusFilter = 'All';
      currentPage = 1;
    });
  }

  void _loadData() {
    setState(() {
      isLoading = true;
    });
    Future.delayed(const Duration(seconds: 1), () {
      setState(() {
        isLoading = false;
      });
    });
  }
}
